mayastor/src/rebuild.rs

```diff
@@ -1,284 +1,224 @@
-        } else {
-            self.state = RebuildState::Failed;
+        let segment_size = 10 * 1024;
+        // validation passed, block size is the same for both
+        let block_size = destination_hdl.get_bdev().block_len() as u64;
+        let segment_size_blks = (segment_size / block_size) as u64;
+
+        let copy_buffer = source_hdl
+            .dma_malloc((segment_size_blks * block_size) as usize)
+            .context(NoCopyBuffer {})?;
+
+        Ok(RebuildTask {
+            nexus_name,
+            source,
+            source_hdl,
+            destination,
+            destination_hdl,
+            start,
+            end,
+            current: start,
+            block_size,
+            segment_size_blks,
+            copy_buffer,
+            complete,
+            state: RebuildState::Pending,
+        })
+    }
+
+    /// rebuild a non-healthy child from a healthy child from start to end
+    async fn run(&mut self) {
+        self.state = RebuildState::Running;
+        self.current = self.start;
+        self.stats();
+
+        while self.current < self.end {
+            if let Err(e) = self.copy_one().await {
+                error!("Failed to copy segment {}", e);
+                self.state = RebuildState::Failed;
+                self.send_complete();
+            }
+            // TODO: check if the task received a "pause/stop" request, eg child
+            // is being removed
         }
 
-        let _ = self.progress.take();
-        self.send_completion(self.state);
+        self.state = RebuildState::Completed;
+        self.send_complete();
     }
 
-    /// determine the next segment for which we will issue a rebuild
-    #[inline]
-    fn num_blocks(&mut self) -> u32 {
-        if self.num_segments > 0 {
-            self.blocks_per_segment
-        } else {
-            self.num_segments += 1;
-            self.partial_segment
-        }
-    }
-
-    pub fn current(&self) -> u64 {
-        self.current_lba
-    }
+    /// copy one segment worth of data from source into destination
+    async fn copy_one(&mut self) -> Result<(), RebuildError> {
+        // Adjust size of the last segment
+        if (self.current + self.segment_size_blks) >= self.start + self.end {
+            self.segment_size_blks = self.end - self.current;
 
-    /// Copy blocks from source to target with increments of segment size.
-    /// When the task has been completed, this function returns
-    /// Ok(true). When a new IO has been successfully dispatched in returns
-    /// Ok(false)
-    ///
-    /// When memory allocation fails, it shall return an error no attempts will
-    /// be made to restart a build automatically, ideally we want this to be
-    /// done from the control plane and not internally, but we will implement
-    /// some form of implicit retries.
-    fn next_segment(&mut self) -> Result<RebuildState, Error> {
-        let num_blocks = self.num_blocks();
+            self.copy_buffer = self
```

Comment:
We already allocated a buffer when we constructed the rebuild tasks, why allocate a new one?

Reply:
It's done temporarily as the BDevHandle seem to use the full len of the DmaBuff when issue IO, and there seems to be no way of setting the len to a smaller size.
So we add new read/write fn or enhance the DmaBuff to expose a "virtual" len? hmm